### ISSUE TITLE
Support for User initiated Change Callbacks

### ISSUE THREAD

**Issue by mahaloz** [NationalSecurityAgency/ghidra#3997]
Hi Devs,

Thanks so much for your continued work on a great decompiler! I wanted to ask if Ghidra provided a way to access changes as they happen by users. As an example, say a user renames `main` to `dank_main`, this should trigger some kind of callback like: `function_changed(old_function, new_function)`... 

This feature is present in IDA Pro as the IDB HOOKs and in Binary Ninja as the BinaryNotification class. If it's present in Ghidra would you mind pointing me in the right direction to the code? 

Thanks,
Zion

**Comment by astrelsky** [comment]
I think you are looking for `DockingAction` and will assume that you are building a plugin. These are typically added to a `ComponentProvider` via `addLocalAction`. You can obtain a `ComponentProvider` through the `PluginTool` provided to your plugin in it's contructor using `PluginTool::getComponentProvider` you will just need to know the name of the `ComponentProvider`.

Here is your example as an example:
[EditNameAction](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/function/EditNameAction.java)
Adding actions may be seen here in [FunctionPlugin](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/function/FunctionPlugin.java#L207)

**Comment by mahaloz** [comment]
Hi @astrelsky, thanks for the reply. This is not exactly what I was looking for, as you noted. I'm looking for something that abstractly gets called every time an object is about to change. In this example, I would still want the normal EditNameAction to exist, but I would want to know when a user has activated it and made a change. 

Any other suggestions? Again, thank you for the reply.

**Comment by dragonmacher** [comment]
If I understand your request correctly, you can add a listener to the Program to get these notifications:

```
	/**
	 * Adds a listener for this object.
	 * @param dol listener notified when any change occurs to this domain object
	 */
	public void addListener(DomainObjectListener dol);
```

**Comment by mahaloz** [comment]
Hi @dragonmacher, I'm assuming you are talking about the code here right:
https://github.com/NationalSecurityAgency/ghidra/blob/4636795205c2ac1af797c81172dcb7d761addf16/Ghidra/Framework/Project/src/main/java/ghidra/framework/model/DomainObject.java#L144 ?

> you can add a listener to the Program to get these notifications
If I'm understanding this code right, the level of notification is only at the database level, right? Meaning, I only know when the _entire_ state of a Ghidra DB has changed, but not what has changed. I reference the events that are available:

https://github.com/NationalSecurityAgency/ghidra/blob/4636795205c2ac1af797c81172dcb7d761addf16/Ghidra/Framework/Project/src/main/java/ghidra/framework/model/DomainObject.java#L63

@dragonmacher is there any other more fine-grained change detection? Like the ability to assign a listener to a group of objects (like Functions or Structs)? Otherwise, I think I would need to scan every function in the DB on each change to detect what has changed in a function.

Thank you again for the reply :).

**Comment by dragonmacher** [comment]
You will get notified of all program events via that listener.  You can then check to see if the event contains the change you are interested in.   The specific event type should have the old and new function, depending upon the type of event.  The events are defined in the `ghidra.program.util.ChangeManager` interface.  In your example above, it sounds like you are interested in the `DOCR_SYMBOL_RENAMED` event.     To figure out what events the system generates, you can just print all events out and make changes.

Here is a snippet of how the `FunctionGraphPlugin` uses this listener:
```
        @Override
	public void domainObjectChanged(DomainObjectChangedEvent ev) {
		if (!isVisible()) {
			return;
		}

		if (controller.getGraphedFunction() == null) {
			return;
		}

		boolean graphChangedButNotRebuilt = false;
		boolean rebuildGraphOnChanges = isPerformingGraphRebuildOnDataChanges();

		//
		// Note: since we are not looping and we are using 'else if's, order is important!
		// 

		if (ev.containsEvent(DomainObject.DO_OBJECT_RESTORED) ||
			ev.containsEvent(ChangeManager.DOCR_FUNCTION_BODY_CHANGED)) {
			if (graphDataMissing()) {
				controller.clear();
				return; // something really destructive has happened--give up!
			}

			graphChangedButNotRebuilt = !handleObjectRestored(ev, rebuildGraphOnChanges);
		}
		else if (ev.containsEvent(ChangeManager.DOCR_SYMBOL_ADDED) ||
			ev.containsEvent(ChangeManager.DOCR_SYMBOL_REMOVED)) {

			if (currentGraphContainsEventAddress(ev)) {
				graphChangedButNotRebuilt = !handleSymbolAddedRemoved(ev, rebuildGraphOnChanges);
			}
		}
		else if (ev.containsEvent(ChangeManager.DOCR_MEM_REFERENCE_ADDED) ||
			ev.containsEvent(ChangeManager.DOCR_MEM_REFERENCE_REMOVED)) {

			if (currentGraphContainsReferenceChangedEvent(ev)) {
				graphChangedButNotRebuilt = !handleReferenceAddedRemoved(ev, rebuildGraphOnChanges);
			}
		}
		else if (ev.containsEvent(ChangeManager.DOCR_SYMBOL_RENAMED)) {
			handleSymbolRenamed(ev);
		}

		if (graphChangedButNotRebuilt) {
			updateGraphForAffectedAddresses(ev);
		}

		// force a repaint for any type of event to make sure we are up-to-date, even for events
		// we are not directly interested in.
		controller.repaint();
	}
```

The important bits there are the check for `if (ev.containsEvent(DomainObject.DO_OBJECT_RESTORED)`, since that implies an undo/redo operation, which means you will not get any of the other normal events.   Otherwise, you can just check for the events you desire.

As you go to process events, the events have new and old values, either of which may be null, depending upon the type of event.  Again, debugging/printing the values will be very helpful.

**Comment by mahaloz** [comment]
For searchable posterity, the source referenced above is here:
https://github.com/NationalSecurityAgency/ghidra/blob/1a1d06b749d4d7e505c493e6f8cad34ecf3df28f/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/util/ChangeManager.java#L26

@dragonmacher I dare to say this is **exactly** what I was looking for! Thanks so much. This is also a link to some code that demonstrates the use of the `DomainObjectListener` within a plugin, which grabs change events and parses them: 
https://github.com/NationalSecurityAgency/ghidra/blob/da94eb86bd2b89c8b0ab9bd89e9f0dc5a3157055/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/commentwindow/CommentWindowPlugin.java#L105

**TL;DR:**
Ghidra provides an interface for event parsing in `DomainObjectListener` which will allow you to get a `ProgramChangeRecord` from the events it receives allowing you to see the before-and-after effect of a change. 

Thanks so much again!
